app/controllers/sms_controller.rb

```
@@ -1,14 +1,51 @@
-class SmsController < ApplicationController
+class SmsController < BaseController
   # Twilio will call various endpoints in this controller
-  skip_before_filter :verify_authenticity_token
+
+  rescue_from 'QuestionValidationError' do |e|
+    render json: format_error(e), status: 500
+  end
 
   def received_user_response
     # Twilio calls this endpoint when a user sends a message to us
-    user_response = params[:Body]
-    from          = params[:From]
-    # TODO(Chris): Implement replying logic here
-    puts user_response
-    puts from
+    answer = params[:Body]
+    phone_number = params[:From]
+
+    respondent = Respondent.find_by_phone_number(phone_number.phony_normalized)
+    state = respondent.survey_execution_states.where(
+      status: [SurveyExecutionState.statuses[:initialized], SurveyExecutionState.statuses[:in_progress]]).first
+    question = state.question
+    survey = state.survey
+
+    survey_response = SurveyResponse.find_or_create_by(survey: survey, respondent: respondent)
+    response = Response.create(
+      survey_response: survey_response, respondent: respondent, question: question, answer: answer)
+
+    if question.invalid?(:next_question)
```

This validation shouldn't happen in the sms_controller. It'd be better in the regular model creation flow because we want to return the error to the user when they're creating or editing the survey. If we're in the process of sending the survey out, we should have already built and verified all models and associations.

---

I can add this validation after model generation.